[[Bug 20289]] Improve dictionary sort

bwmilby opened the pull request:
Initial complaint was that "me" is very difficult to locate in the dictionary.  The current priority search ends up with terms in an unsorted state since it does not explicitly maintain the alphabetical sort.  Additionally, the prioritized sort is looking at all syntax terms which is probably more than is needed for the priority portion.

Added a new regex for an exact word match.  This will be the first one used for the sort and will use the display name.  This will allow for things like "me" to appear at the top and other short words to be closer to the top.  The second regex will use the existing "start of word" search, but also only search the display name.  Finally, the display names will be compared to maintain alphabetical order.

bwmilby committed:
[[Bug 20289]] Improve dictionary sort

Initial complaint was that "me" is very difficult to locate in the dictionary.  The current priority search ends up with terms in an unsorted state since it does not explicitly maintain the alphabetical sort.  Additionally, the prioritized sort is looking at all syntax terms which is probably more than is needed for the priority portion.

Added a new regex for an exact word match.  This will be the first one used for the sort and will use the display name.  This will allow for things like "me" to appear at the top and other short words to be closer to the top.  The second regex will use the existing "start of word" search, but also only search the display name.  Finally, the display names will be compared to maintain alphabetical order.

montegoulding commented on Documentation/html_viewer/js/dictionary_functions.js:
```diff
@@ -85,20 +85,24 @@
+        {
+            matchExp += '(?=.*' + matchToken + ')';
+            priorityMatch += '(?=.*\\b' + matchToken + ')';
+            wordMatch += '(?=.*\\b' + matchToken + '\\b)';
```
@bwmilby @livecodeali I'm wondering if this should be tweaked so that the sort order is:

1. whole match of display name
2. display name begins with token
3. display name contains token
4. syntax contains token somewhere

bwmilby replied:
That would be a very easy thing to add.  I'm not sure if we should add 3 or replace 2 with 3.  I'm also wondering how this is impacting the speed of the sort.  It seems faster, but I want to do some actual tests and compare the original to the update (could even do a/b/c with above as c).  I need to figure out the best way to do that - probably with a stand alone browser.

livecodeali replied:
I think we might need to keep 3. because of LiveCode Builder's habit of having display names that are distinct from the syntax.

bwmilby commented on Documentation/html_viewer/js/dictionary_functions.js:
```diff
@@ -108,19 +112,35 @@
+                return -1;
+            if (tPriorityB && !tPriorityA)
+                return 1;
+
```
This is where the addition would be made to include 3. above.  I would probably want to change to using tMatchA & tMatchB for each of the 3 sets of Regex.  The third would use regex.test(tName?).  This will slow things down some, but I have not had a chance to build a page with timers to check all of the performance variations.

montegoulding replied:
@bwmilby The alphabetical sort would mean that wordRegex isn't necessary wouldn't it or is it faster to return that here?

bwmilby replied:
wordRegex will ensure an exact match that isn't the first word appears at the top of the list.

bwmilby replied:
The wordRegex groups whole word matches at the top of the list so "pane" will have the 2 entries that match ahead of the ones with "panel" for example.  Without wordRegex, they are not separated and are just sorted alphabetically.

I'm trying to envision where adding the extra regex will help find things.  "the" is an example where it would change the order, but I'm not sure if it is helpful.  After "there" matches, you either get "dither" or "abs" as your next match.  Adding the extra regex will put matches in the middle/end of a word in the display name ahead of ones that have that kind of match only in the syntax.  I have the change stashed and can submit.

montegoulding replied:
sorry... I was being stupid and thinking the word and priority matches were for the beginning of the display name not anywhere in it

bwmilby committed:
[[ Bug 20289 ]] Add term match to sort function

Per an earlier suggestion, added another regex to the sort function.  Before the alpha sort, added the regex to put term matches within the display name ahead of entries that just had term matches within the syntax.  When checking the letter "t", it increased the search time from around 50ms to 68ms.

Updated sort function to use a single pair of variables for the regex tests and declared them at the top of the function.  This was done to make it easier to add/subtract regex tests in the future.

Updated the spacing in the code to be consistent with the rest of the file.  In this file, tab characters are used vice the normal 4 spaces for indentation that are specified in the style guide.

Found a console.log entry that had not been commented out and corrected it.

livecodeali commented:
@livecode-vulcan review ok ebe217a

livecode-vulcan commented:
💙 review by @livecodeali ok ebe217a7bf74638927ea3e330428091ca01012fe
